File: 2021/2021W48/preppindataW48.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)

final <- read_excel("C:/Data/PreppinData/PD 2021 Wk 48 Input.xlsx", col_names = FALSE) %>%
  drop_na() %>%
  `colnames<-`(c('Measure','Val1','Val2')) %>%
  mutate('Branch' = if_else(str_detect(`Val1`,'([A-z])'),`Measure`, NA_character_)) %>%
  fill(`Branch`, .direction = "down") %>%
  `colnames<-`(c('Measure', .[[1,2]], .[[1,3]], 'Branch')) %>%
  filter(`Branch`!=`Measure`) %>%
  pivot_longer(cols=-c('Measure','Branch'), names_to='Recorded Year', names_pattern='\\D+(\\d+)', values_to='Value') %>%
  mutate('Clean Measure Names' = str_remove(`Measure`,'(\\s\\(.\\))$'),
         'Factor' = paste0(str_extract(`Measure`,'(?<=\\()(.)(?=\\)$)'),''),
         'True Values' = if_else(`Factor`=='m',1000000,if_else(`Factor`=='k',1000,1))*as.numeric(`Value`)) %>%
  select(c('Branch', 'Clean Measure Names', 'Recorded Year', 'True Values'))
  
View(final)

File: 2021/2021W48/PreppindataW48.py
import pandas as pd
import re

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\PD 2021 Wk 48 Input.xlsx")
df = pd.read_excel(xlsx)
df.columns = ['Branch', 'Measure', df[df.columns[2]][0], df[df.columns[3]][0]]

df['Branch'] = df.apply(lambda x: x['Measure'] if isinstance(x[df.columns[2]], str) else None, axis=1)
df['Branch'] = df['Branch'].fillna(method='ffill')
df = df[df['Branch'] != df['Measure']].dropna()

final = df.melt(id_vars=['Branch', 'Measure'], value_name='Value', var_name='Recorded Year')
final['Recorded Year'] = final['Recorded Year'].apply(lambda x: x[-4:]) 
final['Clean Measure Names'] = final['Measure'].apply(lambda x: re.sub('(\s\(.\))$','',x)) 
final['Factor'] = final['Measure'].apply(lambda x: re.search('\((.)\)$',x).group(1) if re.match('.*\(.\)$',x) else '')  
final['True Values'] = final.apply(lambda x: x['Value']*(1000000 if x['Factor']=='m' else 1000 if x['Factor']=='k' else 1), axis=1)
final = final[['Branch', 'Clean Measure Names', 'Recorded Year', 'True Values']]
